cy_ioctl.go

```diff
@@ -730,28 +724,98 @@ func (d *Device) sdpcmProcessRxPacket(buf []byte) (payloadOffset, plen uint32, f
 	case whd.ASYNCEVENT_HEADER:
 		const totalHeaderSize = whd.SDPCM_HEADER_LEN + whd.BDC_HEADER_LEN
 		if hdr.Size <= totalHeaderSize {
-			return 0, 0, badFlag, err8IgnoreTooSmallAsyncPacket
+			return 0, 0, badHeader, err8IgnoreTooSmallAsyncPacket
 		}
 		bdcHeader := whd.DecodeBDCHeader(buf[payloadOffset:])
 		payloadOffset += whd.BDC_HEADER_LEN + uint32(bdcHeader.DataOffset)<<2
 		plen = uint32(hdr.Size) - payloadOffset
 		payload := buf[payloadOffset:]
-		// payload is actually an ethernet packet with type 0x886c.
+		// Check payload is actually an ethernet packet with type 0x886c.
 		if !(payload[12] == 0x88 && payload[13] == 0x6c) {
 			// ethernet packet doesn't have the correct type.
 			// Note - this happens during startup but appears to be expected
-			// return 0, 0, badFlag, Err9WrongPayloadType
-			err = err9WrongPayloadType
+			return 0, 0, badHeader, err9WrongPayloadType
 		}
 		// Check the Broadcom OUI.
 		if !(payload[19] == 0x00 && payload[20] == 0x10 && payload[21] == 0x18) {
-			return 0, 0, badFlag, err10IncorrectOUI
+			return 0, 0, badHeader, err10IncorrectOUI
 		}
 		plen = plen - 24
 		payloadOffset = payloadOffset + 24
 	default:
 		// Unknown Header.
-		return 0, 0, badFlag, err11UnknownHeader
+		return 0, 0, badHeader, err11UnknownHeader
 	}
-	return payloadOffset, plen, headerFlag, err
+	return payloadOffset, plen, headerType, err
+}
+
+// ref: void cyw43_cb_process_async_event(void *cb_data, const cyw43_async_event_t *ev)
+func (d *Device) processAsyncEvent(ev whd.AsyncEvent) error {
+	fmt.Printf("AsyncEvent: %+v\r\n", ev)
+	switch ev.EventType {
+	case whd.CYW43_EV_ESCAN_RESULT:
+		// TODO
+	case whd.CYW43_EV_DISASSOC:
+		go d.notifyDown()
```

Comment:
Does this have to be a goroutine? Worried about processor resources...

Reply:
I was trying to get out from under the Lock.  processAsyncEvent() is called from poll(), which is holding the Lock.  I was worried about the case where the notifyDown call back would call into tcpip and tcpip would try to sendIoctl or Tx, creating a deadlock.  Using a Go func, lets the callback happen with the Lock not held, avoiding a potential deadlock.

Minimally I should put a comment above that line describing the intent.

TinyGo seems to handle Go func overhead pretty good.  Do you think it's going to cause us problems?  It hard to predict the app requirements above us.